diff --exclude=.git -Nur x/BearSSL/inc/bearssl_x509.h BearSSL/inc/bearssl_x509.h
--- x/BearSSL/inc/bearssl_x509.h	2021-11-17 19:01:11.000000000 +0100
+++ BearSSL/inc/bearssl_x509.h	2020-11-12 22:00:49.000000000 +0100
@@ -676,6 +676,9 @@
 	/* The pad serves as destination for various operations. */
 	unsigned char pad[256];
 
+	/* Certificate CN */
+	unsigned char cn[256];
+
 	/* Buffer for EE public key data. */
 	unsigned char ee_pkey_data[BR_X509_BUFSIZE_KEY];
 
diff --exclude=.git -Nur x/BearSSL/src/x509/x509_minimal.c BearSSL/src/x509/x509_minimal.c
--- x/BearSSL/src/x509/x509_minimal.c	2021-11-17 19:01:11.000000000 +0100
+++ BearSSL/src/x509/x509_minimal.c	2020-11-15 16:27:25.000000000 +0100
@@ -1447,6 +1447,9 @@
 
 	size_t n1, n2;
 
+	memcpy(CTX->cn, &CTX->pad[1], CTX->pad[0]);
+	CTX->cn[CTX->cn[0]] = 0;
+
 	if (CTX->server_name == NULL) {
 		T0_PUSH(0);
 		T0_RET();
